File: src/dere_daemon/context_tracking.py
from __future__ import annotations

import re
from typing import Any

from dere_graph.models import EntityEdge, EntityNode


def build_context_metadata(
    nodes: list[EntityNode],
    edges: list[EntityEdge],
) -> dict[str, Any]:
    return {
        "entities": [{"uuid": node.uuid, "name": node.name} for node in nodes],
        "edges": [edge.uuid for edge in edges],
    }


def extract_cited_entity_uuids(
    response_text: str,
    metadata: dict[str, Any] | None,
) -> list[str]:
    if not response_text or not metadata:
        return []

    entities = metadata.get("entities") or []
    if not entities:
        return []

    response_lower = response_text.lower()
    cited: list[str] = []
    seen: set[str] = set()

    for entity in entities:
        uuid = entity.get("uuid")
        name = entity.get("name")
        if not uuid or not name:
            continue
        pattern = r"\b" + re.escape(str(name).lower()) + r"\b"
        if re.search(pattern, response_lower):
            if uuid not in seen:
                cited.append(uuid)
                seen.add(uuid)

    return cited

File: packages/daemon/src/context-tracking.ts
export type ContextMetadata = {
  entities: Array<{ uuid: string; name: string }>;
  edges: string[];
};

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function buildContextMetadata(
  nodes: Array<{ uuid: string; name: string }>,
  edges: Array<{ uuid: string }>,
): ContextMetadata {
  return {
    entities: nodes.map((node) => ({ uuid: node.uuid, name: node.name })),
    edges: edges.map((edge) => edge.uuid),
  };
}

export function extractCitedEntityUuids(
  responseText: string,
  metadata: Record<string, unknown> | null | undefined,
): string[] {
  if (!responseText || !metadata) {
    return [];
  }

  const entities = metadata.entities;
  if (!Array.isArray(entities) || entities.length === 0) {
    return [];
  }

  const responseLower = responseText.toLowerCase();
  const cited: string[] = [];
  const seen = new Set<string>();

  for (const entity of entities) {
    if (!entity || typeof entity !== "object") {
      continue;
    }
    const record = entity as Record<string, unknown>;
    const uuid = typeof record.uuid === "string" ? record.uuid : "";
    const name = typeof record.name === "string" ? record.name : "";
    if (!uuid || !name) {
      continue;
    }
    const pattern = new RegExp(`\\b${escapeRegExp(name.toLowerCase())}\\b`, "i");
    if (pattern.test(responseLower) && !seen.has(uuid)) {
      cited.push(uuid);
      seen.add(uuid);
    }
  }

  return cited;
}
